File: schemas/hr-payroll/drizzle/employees.ts
// employees: Core employee records with personal details, employment type, and status tracking.
// See README.md for full design rationale.

import { pgTable, uuid, text, timestamp, jsonb, index, unique, pgEnum } from "drizzle-orm/pg-core";
import { users } from "./users";

export const employmentTypeEnum = pgEnum("employment_type", [
  "full_time",
  "part_time",
  "contractor",
  "intern",
  "temporary",
]);

export const employeeStatusEnum = pgEnum("employee_status", [
  "active",
  "on_leave",
  "suspended",
  "terminated",
]);

export const employees = pgTable(
  "employees",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    userId: uuid("user_id").references(() => users.id, { onDelete: "set null" }),
    employeeNumber: text("employee_number"),
    firstName: text("first_name").notNull(),
    lastName: text("last_name").notNull(),
    email: text("email").notNull(),
    phone: text("phone"),
    dateOfBirth: text("date_of_birth"),
    hireDate: text("hire_date").notNull(),
    terminationDate: text("termination_date"),
    employmentType: employmentTypeEnum("employment_type").notNull(),
    status: employeeStatusEnum("status").notNull().default("active"),
    metadata: jsonb("metadata"),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
    updatedAt: timestamp("updated_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_employees_employee_number").on(table.employeeNumber),
    index("idx_employees_user_id").on(table.userId),
    index("idx_employees_status").on(table.status),
    index("idx_employees_employment_type").on(table.employmentType),
  ]
);

File: schemas/hr-payroll/spacetimedb/employees.rs
// employees: Core employee record linking identity, employment terms, and status.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

// type: String
#[derive(SpacetimeType, Clone)]
pub enum EmploymentType {
    FullTime,
    PartTime,
    Contractor,
    Intern,
    Temporary,
}

// type: String
#[derive(SpacetimeType, Clone)]
pub enum EmployeeStatus {
    Active,
    OnLeave,
    Suspended,
    Terminated,
}

#[spacetimedb::table(name = employees, public)]
pub struct Employee {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub user_id: Option<String>, // UUID — FK → users.id (set null)

    #[unique]
    pub employee_number: Option<String>,

    pub first_name: String,
    pub last_name: String,
    pub email: String,
    pub phone: Option<String>,
    pub date_of_birth: Option<String>,
    pub hire_date: String,
    pub termination_date: Option<String>,

    #[index(btree)]
    pub employment_type: EmploymentType,

    #[index(btree)]
    pub status: EmployeeStatus, // default: Active

    pub metadata: Option<String>, // JSON

    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
